app/src/citations.py

```diff
@@ -75,6 +76,37 @@ def dereference_citations(
     return citations
 
 
+CitationNumberAndSubstring: TypeAlias = dict[int, str]
+ChunkWithCitation: TypeAlias = dict[Chunk, list[CitationNumberAndSubstring]]
+
+
+def combine_citations_by_document(
+    citations: dict[ChunkWithSubsection, int]
+) -> dict[Document, list[ChunkWithCitation]]:
+    """
+    This function groups the chunks by document and nests the values of the citation number and
+    substring.
+
+    Args:
+        citations (dict[ChunkWithSubsection, int])
+        ChunkWithSubsection is the Chunk item with subsection string
+        and int is the user friendly citation number of that substring chunk
```

Comment:
Let's use `subsection` to avoid adding another synonym.
```suggestion
        and int is the user-friendly citation number of that subsection
```